Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,590 @@
+      exists_prop]
+    refine ⟨z, And.intro (And.intro z_s fun z_x ↦ not_mem_empty y ?_) hz⟩
+    rw [← ball_empt]
+    rw [z_x] at hz
+    exact mem_inter y_F hz
+  apply not_lt_of_le (coverMincard_le_card smaller_cover)
+  rw [← h']
+  exact_mod_cast Finset.card_erase_lt_of_mem x_s
+
+open ENNReal EReal
+
+lemma log_coverMincard_nonneg (T : X → X) {F : Set X} (h : F.Nonempty) (U : Set (X × X)) (n : ℕ) :
+    0 ≤ log (coverMincard T F U n) := by
+  apply zero_le_log_iff.2
+  rw [← ENat.toENNReal_one, ENat.toENNReal_le]
+  exact (coverMincard_pos_iff T F U n).2 h
+
+lemma log_coverMincard_iterate_le {T : X → X} {F : Set X} (F_inv : MapsTo T F F) {U : Set (X × X)}
+    (U_symm : SymmetricRel U) (m : ℕ) {n : ℕ} (n_pos : 0 < n) :
+    log (coverMincard T F (U ○ U) (m * n)) / n ≤ log (coverMincard T F U m) := by
+  apply (EReal.div_le_iff_le_mul (b := n) (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)).2
+  rw [← log_pow, StrictMono.le_iff_le log_strictMono]
+  nth_rw 2 [← ENat.toENNRealRingHom_apply]
+  rw [← RingHom.map_pow ENat.toENNRealRingHom _ n, ENat.toENNRealRingHom_apply, ENat.toENNReal_le]
+  exact coverMincard_iterate_le_pow F_inv U_symm m n
+
+lemma log_coverMincard_comp_le {T : X → X} {F : Set X} (F_inv : MapsTo T F F)
+    {U : Set (X × X)} (U_symm : SymmetricRel U) {m n : ℕ} (m_pos : 0 < m) (n_pos : 0 < n) :
+    log (coverMincard T F (U ○ U) n) / n
+    ≤ log (coverMincard T F U m) / m + log (coverMincard T F U m) / n := by
+  rcases F.eq_empty_or_nonempty with rfl | F_nemp
+  · rw [coverMincard_empty, ENat.toENNReal_zero, log_zero,
+      bot_div_of_pos_ne_top (Nat.cast_pos'.2 n_pos) (natCast_ne_top n)]
+    exact bot_le
+  have h_nm : (0 : EReal) ≤ (n / m : ℕ) := Nat.cast_nonneg' (n / m)
+  have h_log := log_coverMincard_nonneg T F_nemp U m
+  have n_div_n := EReal.div_self (natCast_ne_bot n) (natCast_ne_top n)
+    (ne_of_gt (Nat.cast_pos'.2 n_pos))
+  apply le_trans <| div_le_div_right_of_nonneg (Nat.cast_pos'.2 n_pos).le
+    (log_monotone (ENat.toENNReal_le.2 (coverMincard_comp_le_pow F_inv U_symm m_pos n)))
+  rw [ENat.toENNReal_pow, log_pow, Nat.cast_add, Nat.cast_one, right_distrib_of_nonneg h_nm
+    zero_le_one, one_mul, div_right_distrib_of_nonneg (Left.mul_nonneg h_nm h_log) h_log, mul_comm,
+    ← EReal.mul_div, div_eq_mul_inv _ (m : EReal)]
+  apply add_le_add_right (mul_le_mul_of_nonneg_left _ h_log)
+  apply le_of_le_of_eq <| div_le_div_right_of_nonneg (Nat.cast_pos'.2 n_pos).le (natCast_div_le n m)
+  rw [EReal.div_div, mul_comm, ← EReal.div_div, n_div_n, one_div (m : EReal)]
+
+/-! ### Cover entropy of entourages -/
+
+open Filter
+
+/-- The entropy of an entourage `U`, defined as the exponential rate of growth of the size of the
+  smallest `(n, U)`-refined cover of `F`. Takes values in the space fo extended real numbers
+  `[-∞, +∞]`. This first version uses a `liminf`.-/
+noncomputable def coverEntropyInfUni (T : X → X) (F : Set X) (U : Set (X × X)) :=
```

Comment:
I'm not sure about the name. What is `Uni` supposed to mean?

Reply:
It stands for "entourage", obviously.

I've changed `Uni` for `Ent`. I'm still looking for better ideas, but this object has no standard name that I know of, and this is the best we could come up with.